### DISCUSSION TITLE
iOS: can't get web context although it's  render engine is  WKWebview

### DISCUSSION THREAD

**Discussion by billychen2014** [appium/appium#15456] [Q&A]
i just tested the method of finding element  "find_element_by_accessibility_id" in two different env，one is web and another is hybrid web(not exactly name but can be called hybrid app ), result as below:

----hybrid result-----
![image](https://user-images.githubusercontent.com/7981655/120442144-e4345b00-c3b7-11eb-9a2a-555107a40509.png)

---web result--------
![image](https://user-images.githubusercontent.com/7981655/120440646-57d56880-c3b6-11eb-8f96-35b20e957d9f.png)

we can find that the time Consumption is 7665ms VS 490ms.
then i checked the appium server logs and found that  hybrid app spent  about 7 seconds before getting the status code 200, i am curious what happened in the 7 seconds.

2021-06-02 14:50:45:951 [simctl] WD Proxy Proxying [POST /element] to [POST http://127.0.0.1:8100/session/857EF26E-DF9E-4935-B6F4-5353AD9D21B3/element] with body: {"using":"accessibility id","value":"getNetworkStatus-正常参数"}
2021-06-02 14:50:52:485 [simctl] WD Proxy Got response with status 200: 

any body found the same condition?

**Comment by billychen2014** [comment]
https://gist.github.com/billychen2014/596842d6c163890ae861a0dfe9b456d3
attach the details server logs

**Comment by jlipps** [comment]
You don't want to use accessibility ID for web elements in hybrid apps. iOS promotes web elements to native elements, and there are so many web elements it makes element lookups really slow. To find web elements, switch to the web context and use web locator strategies like CSS selector. This will be much faster.

**Reply by billychen2014** [reply]
thanks.
i tested and found that there is only one context ,i.e.    ['NATIVE_APP'] but  not web context.

i asked help from dev and they confirmed that the UI was rendered by WKWebview (browser on iPhone), so , i'm confused that why there is no web context.

i tested the web app[image 2 above] and found that there also only ['NATIVE_APP'] context, but the method find by id work well, hmm..

**Reply by KazuCocoa** [reply]
Does Web Inspector work against the app under test?
https://webkit.org/web-inspector/enabling-web-inspector/

Getting Web Context on real iOS devices may take a few seconds sometimes.
`appium:webviewConnectTimeout` capability may help to increase such timeout.
https://github.com/appium/appium-xcuitest-driver#mobile-getcontexts also may help.

**Reply by billychen2014** [reply]
Does Web Inspector work against the app under test?
--yes, it works

Getting Web Context on real iOS devices may take a few seconds sometimes.
appium:webviewConnectTimeout capability may help to increase such timeout.
 --- i tested and wait about 25 seconds ,result is that i can not get web context, hmm

**Reply by KazuCocoa** [reply]
Ok, could you share the full log with `appium:safariLogAllCommunication` and `/context` call?
Perhaps I may help to check Appium <--> WebInspector communication.

It also would be nice to try the latest Appium since we've added a few improvements for remote debugger.
(btw, it seems good to create another discussion thread or change this title. It's up to you :))

**Reply by billychen2014** [reply]
update title and  i will spend some time on 'appium:safariLogAllCommunication' as i need to learn to use it , thanks.
